bdk_chain/src/keychain/keychain_txout_index.rs

```diff
@@ -236,42 +252,37 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
     /// ## Panics
     ///
     /// Panics if the `keychain` has not been added to the index.
-    pub fn derive_new(&mut self, keychain: &K) -> (u32, &Script) {
-        let secp = Secp256k1::verification_only();
-        let next_derivation_index = self.next_derivation_index(keychain);
-        let descriptor = self
-            .keychains
-            .get(&keychain)
-            .expect(&format!("no descriptor for keychain {:?}", keychain));
-
-        let new_spk = descriptor
-            .at_derivation_index(next_derivation_index as u32)
-            .derived_descriptor(&secp)
-            .expect("the descriptor cannot need hardened derivation")
-            .script_pubkey();
+    pub fn derive_new(&mut self, keychain: &K) -> (Option<(u32, &Script)>, DerivationAdditions<K>) {
+        let next_derivation_index = match self.next_derivation_index(keychain) {
+            Some(index) => index,
+            None => return (None, DerivationAdditions::default()),
+        };
+        let additions = self.store_up_to(keychain, next_derivation_index);
 
         let index = (keychain.clone(), next_derivation_index);
-        self.inner.insert_script_pubkey(index.clone(), new_spk);
-        let new_spk = self
-            .inner
-            .script_pubkeys()
-            .get(&index)
-            .expect("we just added it");
-        (next_derivation_index, new_spk)
+        let script = self.inner.spk_at_index(&index).expect("we just stored it");
+        (Some((next_derivation_index, script)), additions)
     }
 
     /// Gets the next usued script pubkey in the keychain i.e. the script pubkey with the lowest index that has not been used yet.
     ///
     /// ## Panics
     ///
     /// Panics if `keychain` has never been added to the index
-    pub fn next_unused(&mut self, keychain: &K) -> (u32, &Script) {
+    pub fn next_unused(
+        &mut self,
+        keychain: &K,
+    ) -> (Option<(u32, &Script)>, DerivationAdditions<K>) {
```

Comment:
Same API question as above. What to do when user gets a `None` here?

Reply:
The user knows that there is no more unused. This is excellent information imo.

Reply:
I think that it is useful information but can't we communicate this with an empty changeset?